### PR TITLE
feat: add OOG error handling for `CALLCODE`, `DELEGATECALL` and `STATICCALL` to `ErrorOOGCall`

### PR DESCRIPTION
**Pull request by silathdiir**
Close https://github.com/privacy-scaling-explorations/zkevm-specs/issues/360

Circuit PR https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1127

### Summary

1. Merge `ExecutionState.ErrorOutOfGasCALLCODE`, `ErrorOutOfGasDELEGATECALL` and `ErrorOutOfGasSTATICCALL` into `ExecutionState.ErrorOutOfGasCall`.

2. Update both spec and test code for OOG call.

### CODE REVIEW

File: src/zkevm_specs/evm/execution/oog_call.py [hunk 1]
```diff
@@ -39,18 +45,22 @@ def oog_call(instruction: Instruction):
     is_to_end_tx = instruction.is_equal(instruction.next.execution_state, ExecutionState.EndTx)
     instruction.constrain_equal(FQ(instruction.curr.is_root), is_to_end_tx)
 
+    # Both CALL and CALLCODE opcodes have an extra stack pop `value` relative to
+    # DELEGATECALL and STATICCALL.
+    rw_counter_delta = 11 + is_call + is_callcode
```
Comment (han0110) [inline src/zkevm_specs/evm/execution/oog_call.py:50]: Need to sync the update in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/362, also see this commit for the fix https://github.com/han0110/zkevm-specs/commit/9c4ed00cb29c604a3c6cd89e8b986091eaeef592 which also fixes other testing files.

**Review by han0110 — COMMENTED** [review]
Overall LGTM!

**Review by han0110 — APPROVED** [review]
LGTM!

**Comment by silathdiir** [comment]
Fixed the last step (STOP) of `internal` call test cases as:
```
# Add `callee_reversible_write_counter` to its rw_counter constant to verify reverted.
rw_counter = CONSTANT + callee_reversible_write_counter

# Set its (`caller`) reversible_write_counter.
reversible_write_counter=caller_reversible_write_counter
```

I also update these fixes to previous PR of `ErrorInvalidOpcode` https://github.com/privacy-scaling-explorations/zkevm-specs/pull/362 (not confirm which will be merged first).

Hi @han0110, I have a simple question about `callee_reversible_write_counter = 2`, is it only set for testing? Not the real value of `reversible_write_counter`. Thanks.

**Comment by han0110** [comment]
> I have a simple question about `callee_reversible_write_counter = 2`, is it only set for testing?

Yes, just setting it to non-zero value for testing, and `2` is chosen because `BeginTx` always does 2 reversible write (balance update) for the underlying call (so actually it shouldn't have been `0`).

File: tests/evm/test_oog_call.py [hunk 2]
```diff
@@ -1,127 +1,172 @@
...
+    rw_dictionary \
+        .stack_read(1, 1020, RLC(stack.cd_offset, randomness)) \
+        .stack_read(1, 1021, RLC(stack.cd_length, randomness)) \
+        .stack_read(1, 1022, RLC(stack.rd_offset, randomness)) \
+        .stack_read(1, 1023, RLC(stack.rd_length, randomness)) \
+        .stack_write(1, 1023, RLC(is_success, randomness)) \
+        .account_read(callee.address, AccountFieldTag.CodeHash, callee_bytecode_hash) \
+        .tx_access_list_account_read(1, callee.address, is_warm_access) \
+        .call_context_read(1, CallContextFieldTag.IsSuccess, 0)
+    # fmt on
```
Comment (lispc) [inline tests/evm/test_oog_call.py:170]: fmt on or fmt: on?

Comment (silathdiir) [reply tests/evm/test_oog_call.py:170]: Fixed to `fmt: on` in both line 169 and 260 in [94fe42d](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/368/commits/94fe42d09f76b84ef3fe90cf556b9926046638ae)

**Review by lispc — APPROVED** [review]

**Pull request merged**